File: aat-lib/src/main/java/ch/bailu/aat_lib/service/cache/elevation/ObjTileElevationColor.java
package ch.bailu.aat_lib.service.cache.elevation;

import org.mapsforge.core.model.Tile;

import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.map.tile.MapTileInterface;
import ch.bailu.aat_lib.service.cache.Obj;
import ch.bailu.aat_lib.service.elevation.tile.DemProvider;
import ch.bailu.aat_lib.util.color.AltitudeColorTable;

public final class ObjTileElevationColor extends ObjTileElevation {

    public ObjTileElevationColor(String id, MapTileInterface b, Tile t, int _split) {
        super(id, b, t, _split);
    }

    @Override
    public void fillBuffer(int[] buffer, Raster raster, SubTile subTile, DemProvider dem) {
        final int dim = dem.getDim().DIM;
        final int bitmap_dim = subTile.pixelDim();

        int c=0;
        int old_line=-1;
        int color=0;

        for (int la = subTile.laSpan.firstPixelIndex(); la <= subTile.laSpan.lastPixelIndex(); la++) {

            final int line = raster.toLaRaster[la]*dim;
            int offset = -1;

            if (old_line != line) {


                for (int lo = subTile.loSpan.firstPixelIndex(); lo <= subTile.loSpan.lastPixelIndex(); lo++) {
                    final int new_offset=raster.toLoRaster[lo];

                    if (new_offset != offset) {
                        offset = new_offset;
                        color = AltitudeColorTable.instance().getColor(dem.getElevation(line + offset));
                    }

                    buffer[c]=color;
                    c++;
                }
            } else {
                copyLine(buffer, c-bitmap_dim, c);
                c+=bitmap_dim;
            }

            old_line=line;
        }
    }

    private void copyLine(int[] buffer, int sourceIndex, int destinationIndex) {
        final int nextLine = destinationIndex;

        for (; sourceIndex < nextLine; sourceIndex++) {
            buffer[destinationIndex] = buffer[sourceIndex];
            destinationIndex++;
        }
    }

    public static class Factory extends Obj.Factory {
        private static final int SPLIT=0;
        private final Tile mapTile;

        public Factory(Tile t) {
            mapTile=t;
        }


        @Override
        public Obj factory(String id, AppContext ac) {
            return  new ObjTileElevationColor(id, ac.createMapTile(), mapTile,SPLIT);
        }
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/cache/elevation/ObjTileElevationColor.kt
package ch.bailu.aat_lib.service.cache.elevation

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.lib.color.AltitudeColorTable
import ch.bailu.aat_lib.map.tile.MapTileInterface
import ch.bailu.aat_lib.service.cache.Obj
import ch.bailu.aat_lib.service.elevation.tile.DemProvider
import org.mapsforge.core.model.Tile

class ObjTileElevationColor(id: String, b: MapTileInterface, t: Tile, split: Int) :
    ObjTileElevation(id, b, t, split) {
    override fun fillBuffer(
        bitmap: IntArray,
        raster: Raster,
        span: SubTile,
        demtile: DemProvider
    ) {
        val dim = demtile.dim.DIM
        val bitmapDim = span.pixelDim()

        var c = 0
        var oldLine = -1
        var color = 0

        for (la in span.laSpan.firstPixelIndex()..span.laSpan.lastPixelIndex()) {
            val line = raster.toLaRaster[la] * dim
            var offset = -1

            if (oldLine != line) {
                for (lo in span.loSpan.firstPixelIndex()..span.loSpan.lastPixelIndex()) {
                    val newOffset = raster.toLoRaster[lo]

                    if (newOffset != offset) {
                        offset = newOffset
                        color = AltitudeColorTable.instance()
                            .getColor(demtile.getElevation(line + offset).toInt())
                    }

                    bitmap[c] = color
                    c++
                }
            } else {
                copyLine(bitmap, c - bitmapDim, c)
                c += bitmapDim
            }

            oldLine = line
        }
    }

    private fun copyLine(buffer: IntArray, sourceIndex: Int, destinationIndex: Int) {
        var si = sourceIndex
        var di = destinationIndex
        val nextLine = di

        while (si < nextLine) {
            buffer[di] = buffer[si]
            di++
            si++
        }
    }

    class Factory(private val mapTile: Tile) : Obj.Factory() {
        override fun factory(id: String, appContext: AppContext): Obj {
            return ObjTileElevationColor(id, appContext.createMapTile(), mapTile, SPLIT)
        }

        companion object {
            private const val SPLIT = 0
        }
    }
}
